### DISCUSSION TITLE
ImageMagick apply shadow effect in bulk

### DISCUSSION THREAD

**Discussion by webbooo** [ImageMagick/ImageMagick#5140] [Help] [answered]
I'm trying to apply a drop shadow effect to some files i have in a folder. Similar to whats done in Photoshop but I'm trying to apply in bulk.

![sample](https://user-images.githubusercontent.com/103241119/167388694-ed35c118-a8d8-4eaa-9d94-941fe97e9079.png)

This is what I'm trying to achieve

![sample1](https://user-images.githubusercontent.com/103241119/167388754-c42c1c5c-8822-48fe-8a90-39148b9f2441.png)

I've used mogrify -modulate to darken *.png's in folder but curious if there is a similar option for a shadow effect?

Thanks

**Comment by webbooo** [comment]
I discovered the below script but this turns out as below and also changes the dimensions of the image.  would like to keep at 500x500

`convert -page +4+4 "old.png" -alpha set \( +clone -background navy -shadow 60x8+8+8 \) +swap -background none -mosaic "new.png"`

![VoL4O](https://user-images.githubusercontent.com/103241119/167414641-88024593-7895-4328-a4cc-3a5b8be2d790.png)

Thanks

**Comment by fmw42** [comment, marked as answer]
To make a shadow, you need two images. Mogrify does not allow that. So you would have to write a script FOR loop over your folder of images and do for each:

`convert red_blue.png \( +clone  -background black  -shadow 80x20-20+0 \) +swap -background none -layers merge  +repage red_blue_shadow.png`

![red_blue_shadow](https://user-images.githubusercontent.com/13458218/167442210-3ddb5abb-56bf-4c9a-bfd0-e9ecaa0c522c.png)

See https://legacy.imagemagick.org/Usage/blur/#shadow

**Reply by webbooo** [reply]
Thanks for this.  Unfortunately this didn't work for me.  Even on a single file I achieved just a shadow image

`convert old.png \( +clone  -background black  -shadow 80x20-20+0 \) +swap -background none -layers merge  +repage red_blue_shadow.png`

![red_blue_shadow](https://user-images.githubusercontent.com/103241119/167513156-b18d579c-fada-4e75-9fe4-5aa8d691cc75.png)

These are the errors I'm getting>

convert: unable to open image '\(': No such file or directory @ error/blob.c/OpenBlob/3527.
convert: no decode delegate for this image format `' @ error/constitute.c/ReadImage/737.
convert: unable to open image '\)': No such file or directory @ error/blob.c/OpenBlob/3527.
convert: unable to open image '\)': No such file or directory @ error/blob.c/OpenBlob/3527.
convert: no decode delegate for this image format `' @ error/constitute.c/ReadImage/737.

I'm on Windows 10 using IM 7.1

**Reply by fmw42** [reply]
For IM 7 on Windows the syntax is slightly different. Try

`magick old.png ( +clone -background black -shadow 80x20-20+0 ) +swap -background none -layers merge +repage red_blue_shadow.png`

Always best to identify your IM version and platform/OS when asking questions on this forum.

**Reply by webbooo** [reply]
sorry I forgot to include that and often forget others use different OS.

Thank you for this it did help but I noticed the output image has increased. The original was 500x500 and the output is 580x580.

Is there a way to keep the image size?  I'm not concerned if the part of the shadow that would be on the edge is cropped.

**Reply by webbooo** [reply]
as an example, here's the difference between the original and the output

![sample1](https://user-images.githubusercontent.com/103241119/167517922-898df8fa-f1a3-446d-bbf4-a7cf8b965922.png)

![sample2](https://user-images.githubusercontent.com/103241119/167517931-aebfa3ac-06a5-48a9-bc7b-5fd3636525c6.png)

appreciate your help

**Reply by fmw42** [reply]
Other than post-cropping, I do not know of any way. You can add a crop to the command, but it can cut off some shadow. The image is expanded automatically, internally to the command to permit shadows on the edges of images.

**Reply by webbooo** [reply]
Thanks, bit of a bugger as while the canvas is 500x500 most of that which are having the shadow applied are centered with much room around the objects so its unfortunate it has to extend the canvas as it seems to have created space for no reason although I understand how it applies the shadow.  I'll have a look at cropping and see if that can give me something that will work.

**Reply by fmw42** [reply]
Try this:

`magick red_blue.png -set option:dims "%wx%h" ( +clone -background black -shadow 80x20-20+0 ) +swap -background none -layers merge +repage -gravity center -crop "%[dims]+20+0" +repage red_blue_shadow.png`

![red_blue_shadow](https://user-images.githubusercontent.com/13458218/167544532-ab5887d6-292c-4f56-aa70-d60d500e1322.png)

For other ways, see https://legacy.imagemagick.org/Usage/blur/#shadow_offset

As I am not a Windows user, my syntax may be wrong. You may have to escape the %. A Windows user may have to help you if my command does not work.  (The unix version did work and produced the example result).

**Reply by webbooo** [reply]
thanks for this fmw42, this worked well on a single file but I haven't been able to get it to do each file in the folder or Loop

This is the batch file I created (i bit of trial and error)
```

set NUM=0

for %%F in (I:\source\*.png) do magick %%F I:\source\%%~nF.png -set option:dims "%wx%h" ( +clone -background black -shadow 80x20-20+0 ) +swap -background none -layers merge +repage -gravity center -crop "%[dims]+20+0" +repage I:\outcome\%%~nF.png
```

I'm getting this error when trying to run it.

magick: MissingArgument `-crop' at CLI arg 21 @ fatal/magick-cli.c/ProcessCommandOptions/447.

again, thanks for your help

**Reply by fmw42** [reply]
Sorry, I do not work on Windows, so do not know .bat file scripting. One of the Windows users will have to help you further.

**Comment by snibgo** [comment]
What is your platform? As you have doubled some of the percent signs (%%), I assume you use Windows BAT.

If so, you need to double the other % signs: "%%wx%%h" and "%%[dims]+20+0"

**Reply by webbooo** [reply]
This helped achieve what I needed.  Thanks fmw42 for helping get what I needed and snibgo for the final touches.  I've seen both you 2 on many forums helping others.  Really appreciated.  Thanks very much!
